model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java

```diff
@@ -0,0 +1,286 @@
+        }
+
+        MapGroupEntity entity = new MapGroupEntity(entityId, realm.getId());
+        entity.setName(name);
+        entity.setParentId(toParent == null ? null : toParent.getId());
+        if (tx.get(entity.getId(), groupStore::get) != null) {
+            throw new ModelDuplicateException("Group exists: " + entityId);
+        }
+        tx.putIfAbsent(entity.getId(), entity);
+
+        return entityToAdapterFunc(realm).apply(entity);
+    }
+
+    @Override
+    public boolean removeGroup(RealmModel realm, GroupModel group) {
+        LOG.tracef("removeGroup(%s, %s)%s", realm, group, getShortStackTrace());
+        if (group == null) return false;
+
+        // TODO: Sending an event (, user group removal and realm default groups) should be extracted to store layer
+        session.getKeycloakSessionFactory().publish(new GroupModel.GroupRemovedEvent() {
+
+            @Override
+            public RealmModel getRealm() {
+                return realm;
+            }
+
+            @Override
+            public GroupModel getGroup() {
+                return group;
+            }
+
+            @Override
+            public KeycloakSession getKeycloakSession() {
+                return session;
+            }
+        });
+
+        session.users().preRemove(realm, group);
+        realm.removeDefaultGroup(group);
+
+        group.getSubGroupsStream().forEach(subGroup -> session.groups().removeGroup(realm, subGroup));
+
+        // TODO: ^^^^^^^ Up to here
+
+        tx.remove(UUID.fromString(group.getId()));
+        
+        return true;
+    }
+
+    /* TODO: investigate following two methods, it seems they could be moved to model layer */
+
+    @Override
+    public void moveGroup(RealmModel realm, GroupModel group, GroupModel toParent) {
+        LOG.tracef("moveGroup(%s, %s, %s)%s", realm, group, toParent, getShortStackTrace());
+        if (toParent != null && group.getId().equals(toParent.getId())) {
+            return;
+        }
+        if (group.getParentId() != null) {
+            group.getParent().removeChild(group);
+        }
+        group.setParent(toParent);
+        if (toParent != null) toParent.addChild(group);
+        else session.groups().addTopLevelGroup(realm, group);
+    }
+
+    @Override
+    public void addTopLevelGroup(RealmModel realm, GroupModel subGroup) {
+        LOG.tracef("addTopLevelGroup(%s, %s)%s", realm, subGroup, getShortStackTrace());
+        subGroup.setParent(null);
+    }
+
+    @Override
+    public void preRemove(RealmModel realm, RoleModel role) {
+        LOG.tracef("preRemove(%s, %s)%s", realm, role, getShortStackTrace());
+        getUnsortedGroupEntitiesStream(realm)
+                .filter(groupEntity -> groupEntity.getGrantedRoles().contains(role.getId()))
+                .map(groupEntity -> session.groups().getGroupById(realm, groupEntity.getId().toString()))
```

Comment:
Extract `groupEntity.getId().toString()`

Reply:
Do you mean to add another `map` function call, with changing `groupEntity` stream to String stream?

Reply:
Ah, sorry, ignore this one. :-)